src/test/java/com/rybka/service/exchange/ExchangeServiceTest.java

```diff
@@ -0,0 +1,157 @@
+        var response = constructResponse(currencyBase, Map.of(currencyTarget, currencyRate));
+
+        when(exchangeRateConnector.retrieveRates(currencyBase)).thenReturn(response);
+
+        // when
+        var actualResult = exchangeService.loadCurrencyOf(currencyBase, currencyTarget);
+
+        // then
+        assertEquals(currencyRate, actualResult.getRate());
+    }
+
+    @Test
+    public void testOnIncorrectCurrency() {
+
+        // given
+        var currencyTarget = "EUR";
+        var incorrectCurrencyRate = 0.0;
+        var incorrectCurrencyTarget = "UAH";
+        var response = constructResponse(incorrectCurrencyTarget, Map.of(currencyTarget, incorrectCurrencyRate));
+
+        when(exchangeRateConnector.retrieveRates(incorrectCurrencyTarget)).thenReturn(response);
+
+        // when
+        var actualResult = exchangeService.loadCurrencyOf(incorrectCurrencyTarget, currencyTarget);
+
+        // then
+        assertEquals(incorrectCurrencyRate, actualResult.getRate());
+    }
+
+    @Test
+    public void testOnCalculation() {
+
+        // given
+        var currencyRate = 0.89;
+        var currencyAmount = 4d;
+
+        // when
+        var expectedResult = currencyRate * currencyAmount;
+        var actualResult = exchangeService.calculateTotal(currencyRate, currencyAmount);
+
+        // then
+        assertEquals(expectedResult, actualResult);
+    }
+
+    @Test
+    public void testOnRetrievingTopCurrency() {
+
+        // given
+        var currencyAmount = 4d;
+        var currencyBase = "USD";
+        var exchangeType = "SELL";
+        var currencyTarget = "EUR";
+        var currencyRate = 0.89;
+        var topCurrencies = "UAH,USD,EUR,PLN";
+        var response = constructResponse(currencyBase, Map.of(currencyTarget, currencyRate));
+        Map<String, BiFunction<String, Double, List<TopCurrencyData>>> exchangeTypeMap = new HashMap<>();
+
+        exchangeService.setExchangeTypeMap(exchangeTypeMap);
+        exchangeService.initExchangeTypeMap();
+
+        when(exchangeProperty.getCurrency()).thenReturn(topCurrencies);
+        when(exchangeRateConnector.retrieveRates(currencyBase)).thenReturn(response);
+
+        // when
+        var actualResult = exchangeService.retrieveTopCurrency(currencyBase, exchangeType, currencyAmount);
+
+        // then
+        assertEquals(currencyAmount, actualResult.getTopCurrencyDataList().size());
+    }
+
+    @Test
+    public void testOnBuyingTopCurrency() {
+
+        // given
+        var currencyAmount = 4d;
+        var currencyBase = "USD";
+        var exchangeType = "BUY";
+        var currencyTarget = "EUR";
+        var currencyRate = 0.89;
+        var topCurrencies = "UAH,USD,EUR,PLN";
+        var response = constructResponse(currencyBase, Map.of(currencyTarget, currencyRate));
+        Map<String, BiFunction<String, Double, List<TopCurrencyData>>> exchangeTypeMap = new HashMap<>();
+
+        exchangeService.setExchangeTypeMap(exchangeTypeMap);
```

Comment:
Map isn't initialized in the class. More it shouldn't be initilized in the test. 
That weird you don't get NPE when run the app, because map isn't initialized but we call .put() method on null pointer.

Reply:
fixed